From 192e4a1d7a04077fe3e94d6eff3ebbd187aa8c05 Mon Sep 17 00:00:00 2001
From: Kevin Harwell <kharwell@digium.com>
Date: Thu, 20 Nov 2014 15:42:01 +0000
Subject: Permission escalation through ConfBridge actions/dialplan functions
CVE: CVE-2014-8417
Origin: http://svnview.digium.com/svn/asterisk?view=rev&rev=428332
Bug: https://issues.asterisk.org/jira/browse/ASTERISK-24490

The CONFBRIDGE dialplan function when executed from an external protocol
(for instance AMI - the Asterisk Manager Interface), could result in a
privilege escalation. Also, the AMI action “ConfbridgeStartRecord” could
also be used to execute arbitrary system commands without first checking
for system access.

Asterisk now inhibits the CONFBRIDGE function from being executed from an
external interface if the live_dangerously option is set to no.  Also, the
“ConfbridgeStartRecord” AMI action is now only allowed to execute under a
user with system level access.

In the default configuration the manager interface is only accessible
from localhost with no users configured and no AGI script is installed,
however using such interfaces is very common.


See Also: http://downloads.asterisk.org/pub/security/AST-2014-017.html
---
 apps/app_confbridge.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/apps/app_confbridge.c b/apps/app_confbridge.c
index 70844d6..2fa7904 100644
--- a/apps/app_confbridge.c
+++ b/apps/app_confbridge.c
@@ -3189,7 +3189,7 @@ static int load_module(void)
 		ast_log(LOG_ERROR, "Unable to load config. Not loading module.\n");
 		return AST_MODULE_LOAD_DECLINE;
 	}
-	if ((ast_custom_function_register(&confbridge_function))) {
+	if ((ast_custom_function_register_escalating(&confbridge_function, AST_CFE_WRITE))) {
 		return AST_MODULE_LOAD_FAILURE;
 	}
 	if ((ast_custom_function_register(&confbridge_info_function))) {
@@ -3220,7 +3220,7 @@ static int load_module(void)
 	res |= ast_manager_register_xml("ConfbridgeKick", EVENT_FLAG_CALL, action_confbridgekick);
 	res |= ast_manager_register_xml("ConfbridgeUnlock", EVENT_FLAG_CALL, action_confbridgeunlock);
 	res |= ast_manager_register_xml("ConfbridgeLock", EVENT_FLAG_CALL, action_confbridgelock);
-	res |= ast_manager_register_xml("ConfbridgeStartRecord", EVENT_FLAG_CALL, action_confbridgestartrecord);
+	res |= ast_manager_register_xml("ConfbridgeStartRecord", EVENT_FLAG_SYSTEM, action_confbridgestartrecord);
 	res |= ast_manager_register_xml("ConfbridgeStopRecord", EVENT_FLAG_CALL, action_confbridgestoprecord);
 	res |= ast_manager_register_xml("ConfbridgeSetSingleVideoSrc", EVENT_FLAG_CALL, action_confbridgesetsinglevideosrc);
 	if (res) {
-- 
2.1.3

